Playwright: refactor EditorPublishPanelComponent out of GutenbergEditorPage.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR refactors the publish checklist panel (known as EditorPublishPanelComponent) out of GutenbergEditorPage.

Related, the return type of `EditorPublishPanelComponent.getPublishedURL` and `GutenbergEditorPage.getPublishedURLFromToast` has been changed to a `URL` type.

Key changes:
- move methods and selectors associated with the publish checklist into new component.
- update specs to call the new methods as necessary.

#### Testing instructions

Ensure that all following builds are green:
  - [x] WPCOM E2E (desktop)
  - [x] WPCOM E2E (mobile)
  - [x] Calypso E2E (desktop)
  - [x] Calypso E2E (mobile)
  - [x] Pre-Release

Related to #60868.

worldomonation committed:
Initial refactor for EditorPublishPanelComponent.
- replace some contents of the `publish` method in GutenbergEditorPage with extracted class methods.
- make sure the test continues to work.

worldomonation committed:
Remove methods and selectors from GutenbergEditorPage.

worldomonation committed:
Transition to use frameLocator instead of passing the frame itself.

worldomonation committed:
Revamp:
- consolidate and update selectors.
- merge the `GutenbergEditorPage` methods `publish` and `update` methods into one.
- change return type of the `publish` and `getPublishedURL` methods to return a URL object.

worldomonation committed:
Update selector in `EditorPublishPanelComponent.publish` to use both `Publish` and `Schedule` selectors.

worldomonation commented on packages/calypso-e2e/src/lib/components/editor-publish-panel-component.ts:
```diff
@@ -0,0 +1,90 @@
+ */
+export class EditorPublishPanelComponent {
+	private page: Page;
+	private frameLocator: FrameLocator;
```
Note the use of `FrameLocator`. This is a [new API](https://playwright.dev/docs/1.18/api/class-framelocator) introduced in Playwright 1.17.

It is the iFrame specific counterpart to the `Locator` API and allows us to locate and interact with elements within an iFrame as if we have the actual frame itself.

Previously:
```
const frameHandle = await this.page.waitForSelector('iframe');
const frame = await frameHandle.contentFrame() as Frame;
await frame.click('selector');
```

Now:
```
const locator = this.frameLocator.locator('selector');
await locator.click()
```

dpasque replied:
Great idea!!

worldomonation commented on packages/calypso-e2e/src/lib/components/editor-publish-panel-component.ts:
```diff
@@ -0,0 +1,90 @@
+	 * The selector used for each state differ; however both have the same end result,
+	 * that is to dismiss the component represented by EditorPublishPanelComponent.
+	 */
+	async closePanel(): Promise< void > {
+		// Construct a comma-separated list of CSS selectors so that one of them will match.
+		const selector = `${ selectors.prePublishButton( 'Cancel' ) }, ${
+			selectors.postPublishClosePanelButton
+		}`;
+		const locator = this.frameLocator.locator( selector );
+		await locator.click();
+	}
```
The idea here is to have a single, obvious method to close the panel, regardless of what the panel is actually showing. The way to dismiss the panel differs depending on where in the publish checklist stage the user is in - but by using a list of CSS selectors, we can simply match on one.

This ethos is reflected throughout other methods and is an attempt to make the API cleaner, less verbose and more intuitive for test writers.

worldomonation commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -41,16 +40,8 @@ const selectors = {
-	addNewButton: '.editor-post-publish-panel a:text-matches("Add a New P(ost|age)")',
-	closePublishPanel: 'button[aria-label="Close panel"]',
+	// Toast
+	toastViewPostLink: '.components-snackbar__content a:text-matches("View (Post|Page)", "i")',
```
Now that the `EditorPublishPanelComponent` has taken over interactions with the `View Post/Page` button that appears after the article is published, the main `GutenbergEditorPage` should only handle the toast, hence the rename and narrowing of the selector.

Plus, without the narrowing, the `Locator` API complains that it has matched on multiple elements.

worldomonation commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -400,40 +396,39 @@ export class GutenbergEditorPage {
-
-		if ( visit ) {
-			await this.visitPublishedPost( publishedURL );
+		if ( update ) {
```
The `update` method has been deprecated and unified into `publish` method, since the contents of both the `publish` and `update` methods were nearly identical. The only difference was the text on the target button.

In the future, this interaction with the toolbar is likely to be refactored into its own component.

dpasque replied:
Hmm, merging the update action into this method through a boolean parameter feels a little unintuitive to me semantically.  Calling `publish( { update: true } )` as the way to update a post (and the required way to to do it) seems like it might initially confuse callers. I think the confusion comes from mixing the syntax for actions that otherwise feeling similar (i.e., two actions are just auto-included by combined selector, but the third similar action is governed by a parameter).

I think we might benefit from taking this in one of two directions:
1. Make every distinct action (update, schedule, publish) its own public method. We can avoid code duplication by modularizing shared pieces into private class methods.
2. Just combine all three actions equally here without the boolean parameter. So this method will basically click the blue button in the top right of the editor no matter what it is, and handle whatever needs to be handled there.

What do you think?

worldomonation replied:
Hmm, good point. I will brood over this and evaluate both options.

worldomonation replied:
I've chosen Path 2 and unified the `publish` and `update` methods into one in 7df6ee16bfd28a4066cd09597dc5522f8a35f093.

worldomonation commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -400,40 +396,39 @@ export class GutenbergEditorPage {
+		// eg. publishing a post with some of the Jetpack Earn blocks.
+		// By racing the two methods of obtaining the published article's URL, we can
+		// guarantee that one or the other works.
+		const publishedURL: URL = await Promise.race( [
+			this.editorPublishPanelComponent.getPublishedURL(),
+			this.getPublishedURLFromToast(),
+		] );
```
As noted in the comment - our editor sometimes has a peculiar issue such as https://github.com/Automattic/wp-calypso/issues/50302 where the EditorPublishPanelComponent is either not present, or forcibly dismissed when publishing/updating/scheduling a post. By racing the promises we can obtain the published article URL even in that case.

worldomonation commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -466,31 +461,12 @@ export class GutenbergEditorPage {
-	 * @param options will be forwarded to the button click action
-	 */
-	async postPublishAddNewItem( options: ClickOptions = {} ): Promise< void > {
-		const frame = await this.getEditorFrame();
-		await frame.waitForSelector( selectors.publishPanel );
-		await frame.click( selectors.addNewButton, options );
+		const toastLocator = frame.locator( selectors.toastViewPostLink );
+		const publishedURL = ( await toastLocator.getAttribute( 'href' ) ) as string;
+		return new URL( publishedURL );
```
This method now only handles the obtaining of published article's URL from the toast.

worldomonation commented on test/e2e/specs/editor/editor__page-basic-flow.ts:
```diff
@@ -26,7 +26,7 @@ describe( DataHelper.createSuiteTitle( 'Editor: Basic Post Flow' ), function ()
 	let pagesPage: PagesPage;
 	let editorSettingsSidebarComponent: EditorSettingsSidebarComponent;
-	let publishedUrl: string;
+	let publishedUrl: URL;
```
Everything after this point are the specs being updated to work with the new return value of type `URL`.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-publish-panel-component.ts:
```diff
@@ -0,0 +1,90 @@
+	// Post-publish
+	postPublishClosePanelButton: `${ panel } button[type="button"]:has(svg[aria-hidden="true"])`, // aria-label changes depending on the UI language used.
+	publishedArticleURL: `${ panel } input[readonly]`,
+	addNewArticleButton: `${ panel } .post-publish-panel__postpublish-buttons > a`,
```
I believe this matches on both Add New and the View Post buttons. Can we maybe do this by `href`?

worldomonation replied:
Good catch, it does match both View Post/Page and Add New Post.

`${ panel } .post-publish-panel__postpublish-buttons a[href="post-new.php?post_type=post"]` should match the Add New Button without relying on the string on the button.

worldomonation replied:
Fixed in 23018cb06a811d773c8ff3036cba8d2a24986e2b.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-publish-panel-component.ts:
```diff
@@ -0,0 +1,90 @@
+		`${ panel } .editor-post-publish-panel__header button:text("${ buttonText }")`,
+
+	// Post-publish
+	postPublishClosePanelButton: `${ panel } button[type="button"]:has(svg[aria-hidden="true"])`, // aria-label changes depending on the UI language used.
```
Just to clarify: do we need to support multiple languages here? I ask because I saw we use the button text in other selectors on this component.

This is definitely an unfortunate selector to have to use and would be nice to avoid, but if we can't use anything with text, I think you're right that we don't have a better option 😬

worldomonation replied:
Ideally, we develop all components going forward with an eye towards i18n, so that means either:
1. accepting the string to be matched;
2. use selectors that do not rely on UI text;

I've struggled with committing to this, but I plan to review the selectors I've implemented here to see if I can discontinue use of text.

worldomonation replied:
I've replaced the whole lot of selectors with non-text based ones.

dpasque replied:
>use selectors that do not rely on UI text;

We might want to spin up a separate thread or P2 to discuss this, but i'm a little hesitant about this as a commitment going forward... This seems to go against our recommendation of using selectors that are user facing. I think our use of text and aria based selectors is one of the things that has made this Playwright transition a lot cleaner than our old Selenium scripts.

What do you think about allowing text-based selecting by default, and then only changing what we need to for I18N compatibility at the time that we need it? I want to avoid lowering the quality of all of our selectors for a potential future use case that may not come.

All that said, I think that the editor is likely something that needs I18N support, and I think what you have here in this PR is just fine, I just wanted to throw that out there for future consideration!

worldomonation replied:
Sure, let's spin up a new P2 and involve the I18N folks for some input on this.

In your subsequent paragraphs you did cover my reason for going against the Playwright team's recommendation. Notably, that `aria` attributes and UI text are subject to change depending on the language selected. This would definitely lead to issues when these methods are intended to run against non-US English interfaces, as we've found out already in https://github.com/Automattic/wp-calypso/pull/60584. 

I didn't make the decision lightly, as you say - it does run counter to what we've done so far.

worldomonation replied:
Drafted a post and requested a quick look from you before I publish.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-publish-panel-component.ts:
```diff
@@ -0,0 +1,90 @@
+	 *
+	 * @param {ClickOptions} options Options to be forwarded to Frame.click().
+	 */
+	async addNewArticle( options: ClickOptions = {} ): Promise< void > {
```
Could we use this refactor as a chance to redo the typing here so that we are decoupled from the Playwright `click` options typings?

In general, I think we should try to have the signatures of the POM class methods be decoupled from the underlying Playwright API. This prevents changes in the Playwright API over time breaking our spec files, and allows flexibility in the POM classes over time. 

So in this case, if it's just important to be able to start a new article without waiting for the editor to load, we could type `options` as an object with a key called "waitForEditorLoad" or something like that, and then the implementation handles calling the `click` method with the right Playwright options.

worldomonation replied:
I've only kept this signature entirely for the moment to keep compatibility with the Tracks test, but as we discussed in Zoom I feel that we should discontinue the Tracks tests entirely for now. That would mean I can remove this method along with the typing as nothing else uses this call.

worldomonation replied:
Removed in 9a2f2090cb4999019a6391663d137e2b63dbd9d2

dpasque commented on packages/calypso-e2e/src/lib/components/editor-publish-panel-component.ts:
```diff
@@ -0,0 +1,90 @@
+	/**
+	 * Returns the URL of the published article.
+	 *
+	 * @returns {string} URL to the published article.
```
Nit: type update here! 👍

worldomonation replied:
Fixed in 23018cb06a811d773c8ff3036cba8d2a24986e2b.

dpasque reviewed (commented):
Overall looking nice -- great to see the refactor starting!

Have a few design questions, let me know what you think!

worldomonation committed:
- narrow selector for Add New button.
- fix JSDoc type declaration.

worldomonation committed:
- remove use of text string from selectors in EditorPublishPanelComponent.
- update return value in JSDoc.

worldomonation committed:
Unify the `publish`, `schedule` and `update` actions into one.

worldomonation commented on packages/calypso-e2e/src/lib/components/editor-publish-panel-component.ts:
```diff
@@ -0,0 +1,90 @@
+		const publishButtonLocator = this.frameLocator.locator( selectors.publishButton );
+		// If the publish button in the before publish panel is not found, then
+		// perform nothing.
+		if ( ( await publishButtonLocator.count() ) === 0 ) {
+			return;
+		}
```
This part had to be added in order to support the unified `publish`/`update` interface.

For updating an article, only the first step of the Publish button needs to be clicked. When this is clicked, the article is updated and then a toast is displayed.

However, since we've now removed the `update` parameter, this means the actions in `GutenbergEditorPage` will be executed regardless, so the guard needs to be added here to short circuit the second Publish step if the panel is not found.

dpasque replied:
Not required, but is there anyway to pull this logic up to the EditorPage `publish` method instead of having it in here? This feels a little strange to have in the component, I think because it's handling edge cases that don't actually involve the publish panel component (because it's never actually rendered). The edge case handled here is really the responsibility of the EditorPage.

Like, could we instead do a promise race in the EditorPage publish method for the panel publish button appearing and the update toast text appearing? Or find something else to key off of there?

worldomonation replied:
I'll take a look. 

It's a bit hazy, but IIRC the reason I put this logic here is that `GutenbergEditorPage` should call `publish` and it _will be done_. How it's done is up to the component implementing this method to handle.

worldomonation replied:
I've restructured the logic in f8ed7192ff8276955ef4101a9b4afd8959830e82.

Now, in GutenbergEditorPage the presence of the EditorPublishPanelComponent is checked using the new member function `this.editorPublishPanelComponent.panelIsOpen`. 
If true, this means the second-stage `publish` button must be clicked.

worldomonation commented:
Two E2E failures in WPCOM are due to unrelated issue fixed [here](https://github.com/Automattic/wp-calypso/pull/61459).

worldomonation committed:
Remove Editor: Tracks spec and associated infrastructure.

dpasque commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -31,26 +30,18 @@ const selectors = {
+	publishButton: `.editor-post-publish-button__button`,
+	// publishButton: ( parentSelector: string ) =>
+	// `${ parentSelector } button:text("Publish")[aria-disabled=false]`,
 	switchToDraftButton: 'button.editor-post-switch-to-draft',
-	scheduleButton: ( parentSelector: string ) =>
-		`${ parentSelector } button:has-text("Schedule")[aria-disabled=false]`,
+	// scheduleButton: ( parentSelector: string ) =>
+	// 	`${ parentSelector } button:has-text("Schedule")[aria-disabled=false]`,
```
Heads up: probably meant to delete these commented lines?

worldomonation replied:
Yeah, probably. Oops.

worldomonation replied:
Fixed in f8ed7192ff8276955ef4101a9b4afd8959830e82.

dpasque reviewed (approved):
Changes look great! I have a few more questions, but nothing required, so approving!

Thanks for getting this refactor underway! 🎉

worldomonation committed:
Address review comments:
- remove dead selectors
- move the check for EditorPublishPanelComponent up into GutenbergEditorPage.
